Convert hashSets in parallel before merge

jiebinn opened the pull request:
Before merge, if one of the lhs and rhs is singleLevelSet and the other is twoLevelSet, then the SingleLevelSet will call convertToTwoLevel(). The convert process is serial and not in parallel. It will cost lots of cycle before it cosumes all the singleLevelSet.

The idea of the patch is to convert all the singleLevelSets to twoLevelSets in parallel before merge.

I have tested the patch on Intel 2 x 112 vCPUs SPR server with clickbench and latest upstream ClickHouse. Q5 has got a big **264%** performance improvement and 24 queries have got at least 5% performance gain. The overall geomean of 43 queries has gained **7.4%** more than the base code.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This patch will provide a method to deal with all the hashsets in parallel before merge. 

### The detail to define the performance issue and how to resolve it.
First, we have found that there is a performance drop for the Q5 of clickbench as the core count increases.
![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/d9d76d2b-b585-490d-a150-0a96dc473d81)
Then we have collected a pipeline visualization if the max_threads of thread_pool is 80 or 112.
max_threads = 80
![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/6ac5e45c-c98b-48a4-b820-de4097bb71d6)
max_threads = 112
![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/b7155669-fd39-457e-8c7d-e35e0d32b4be)
If the max_threads increase from 80 to 112, the merge stage does not decrease, but merge time is 3.2x more.
when merging two twoLevelHash, there is already an optimization to start thread_pool and merge in parallel. However, when merge one singleLevelHash and one TwoLevelHash, the singleLevelHash has to convert to twoLevelHash. The convert progress is serial.
If there is at least one singleLevelHash and one TwoLevelHash, all the singleLevelHash have to be converted to twoLevelHash before merged with the other twoLevelHash. We could add a new stage before merge called Prepare_Hash_before_Merge, where all the hashset would be processed before merging. All the singleLevelHash would be converted to twoLevelHash in parallel in this stage instead of in serial in merge stage.
with this patch, Q5 has got 2.64x performance improvement on a 2x112 vCPUs system (max_threads = 112).
![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/789f7430-e45d-45a0-8fb5-b0a9c97a25ee)

jiebinn committed:
Merge pull request #2 from ClickHouse/master

lastest upstream master

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #50748 from jiebinn/ClickHouse:UniqExactSet -->
This is an automated comment for commit 38545463aade2f3019b81484d7324f92dcbd4ec2 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50748/38545463aade2f3019b81484d7324f92dcbd4ec2/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

nickitat commented:
pls take a look

```
Jun 11 02:52:18 FAILED: src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o 
Jun 11 02:52:18 /usr/bin/cmake -E __run_co_compile --launcher="prlimit;--as=10000000000;--data=5000000000;--cpu=1000;/usr/bin/sccache" --tidy="/usr/bin/clang-tidy-cache;/usr/bin/clang-tidy-16;--extra-arg-before=--driver-mode=g++" --source=/build/src/Interpreters/examples/two_level_hash_map.cpp -- /usr/bin/clang++-16 --target=x86_64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DANNOYLIB_MULTITHREADED_BUILD -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=10 -DAWS_SDK_VERSION_PATCH=36 -DBOOST_ASIO_HAS_STD_INVOKE_RESULT=1 -DBOOST_ASIO_STANDALONE=1 -DCARES_STATICLIB -DDUMMY_BACKTRACE -DENABLE_ANNOY -DENABLE_MULTITARGET_CODE=1 -DENABLE_QPL_COMPRESSION -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/build/build_docker/includes/configs -I/build/base/glibc-compatibility/memcpy -I/build/src -I/build/build_docker/src -I/build/build_docker/src/Core/include -I/build/base/base/.. -I/build/build_docker/base/base/.. -I/build/contrib/cctz/include -I/build/base/pcg-random/. -I/build/contrib/libfiu/libfiu -I/build/contrib/miniselect/include -I/build/contrib/zstd/lib -I/build/src/Common/mysqlxx/. -I/build/build_docker/rust/skim/include -isystem /build/contrib/llvm-project/libcxx/include -isystem /build/contrib/llvm-project/libcxxabi/include -isystem /build/contrib/libunwind/include -isystem /build/build_docker/contrib/orc/c++/include -isystem /build/contrib/llvm-project/llvm/include -isystem /build/build_docker/contrib/llvm-project/llvm/include -isystem /build/contrib/abseil-cpp -isystem /build/contrib/croaring/cpp -isystem /build/contrib/croaring/include -isystem /build/contrib/sparsehash-c11 -isystem /build/contrib/cityhash102/include -isystem /build/contrib/boost -isystem /build/base/poco/Net/include -isystem /build/base/poco/Foundation/include -isystem /build/base/poco/NetSSL_OpenSSL/include -isystem /build/base/poco/Crypto/include -isystem /build/contrib/boringssl/include -isystem /build/base/poco/Util/include -isystem /build/base/poco/JSON/include -isystem /build/base/poco/XML/include -isystem /build/contrib/replxx/include -isystem /build/contrib/fmtlib-cmake/../fmtlib/include -isystem /build/contrib/magic_enum/include -isystem /build/contrib/double-conversion -isystem /build/contrib/dragonbox/include -isystem /build/contrib/re2 -isystem /build/build_docker/contrib/re2-cmake -isystem /build/contrib/zlib-ng -isystem /build/build_docker/contrib/zlib-ng-cmake -isystem /build/contrib/pdqsort -isystem /build/contrib/xz/src/liblzma/api -isystem /build/contrib/aws/src/aws-cpp-sdk-core/include -isystem /build/build_docker/contrib/aws-cmake/include -isystem /build/contrib/aws/generated/src/aws-cpp-sdk-s3/include -isystem /build/contrib/aws-c-auth/include -isystem /build/contrib/aws-c-common/include -isystem /build/contrib/aws-c-io/include -isystem /build/contrib/aws-crt-cpp/include -isystem /build/contrib/aws-c-mqtt/include -isystem /build/contrib/aws-c-sdkutils/include -isystem /build/contrib/snappy -isystem /build/build_docker/contrib/snappy-cmake -isystem /build/contrib/libbcrypt -isystem /build/contrib/msgpack-c/include -isystem /build/build_docker/contrib/liburing/src/include-compat -isystem /build/build_docker/contrib/liburing/src/include -isystem /build/contrib/liburing/src/include -isystem /build/contrib/fast_float/include -isystem /build/contrib/NuRaft/include -isystem /build/base/poco/MongoDB/include -isystem /build/build_docker/contrib/mariadb-connector-c-cmake/include-public -isystem /build/contrib/mariadb-connector-c/include -isystem /build/contrib/mariadb-connector-c/libmariadb -isystem /build/build_docker/src/Server/grpc_protos -isystem /build/contrib/grpc/include -isystem /build/contrib/c-ares/src/lib -isystem /build/contrib/c-ares/include -isystem /build/contrib/c-ares-cmake/linux -isystem /build/contrib/google-protobuf/src -isystem /build/contrib/s2geometry/src -isystem /build/contrib/AMQP-CPP/include -isystem /build/contrib/AMQP-CPP -isystem /build/contrib/libuv/include -isystem /build/contrib/sqlite-amalgamation -isystem /build/contrib/rocksdb/include -isystem /build/contrib/libpqxx/include -isystem /build/contrib/libpq -isystem /build/contrib/libpq/include -isystem /build/contrib/qpl-cmake -isystem /build/contrib/qpl/include -isystem /build/contrib/idxd-config/accfg -isystem /build/contrib/libstemmer_c/include -isystem /build/contrib/wordnet-blast -isystem /build/contrib/lemmagen-c/include -isystem /build/contrib/ulid-c/include -isystem /build/contrib/consistent-hashing -isystem /build/contrib/annoy/src --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -UNDEBUG -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -g -O0 -g -gdwarf-4  -D_LIBCPP_DEBUG=0 -std=c++2b   -D OS_LINUX -I/build/base -I/build/contrib/magic_enum/include -include /build/src/Core/iostream_debug_helpers.h -Werror -nostdinc++ -MD -MT src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -MF src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o.d -o src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -c /build/src/Interpreters/examples/two_level_hash_map.cpp
Jun 11 02:52:18 /build/src/AggregateFunctions/UniquesHashSet.h:369:91: error: unknown type name 'ThreadPool' [clang-diagnostic-error]
Jun 11 02:52:18     static void parallelizeMergePrepare(const std::vector<UniquesHashSet *> & /*places*/, ThreadPool * /*thread_pool = nullptr*/) {}
```

nickitat commented on src/Interpreters/Aggregator.cpp:
```diff
@@ -2601,6 +2601,21 @@ void NO_INLINE Aggregator::mergeWithoutKeyDataImpl(
 
+    for (size_t i = 0; i < params.aggregates_size; ++i)
+    {
+        if (aggregate_functions[i]->isAbleToParallelizeMerge() &&
```
`isAbleToParallelizeMerge` looks irrelevant here, `isParallelizeMergePrepareNeeded` should be enough

jiebinn replied:
Keep the `isAbleToParallelizeMerge` and delete 'isParallelizeMergePrepareNeeded'.

jiebinn replied:
The flag `isAbleToParallelizeMerge` is for the parallel merge of UniqExact. If the merge stage is not parallel (`isAbleToParallelizeMerge = false`), we could still do hashset converting in parallel. The merge stage and hashset converting stage are independent and irrelevant. Then I think `isAbleToParallelizeMerge` could be removed.

nickitat commented on src/AggregateFunctions/UniqExactSet.h:
```diff
@@ -27,6 +27,56 @@ class UniqExactSet
+    /// In merge, if one of the lhs and rhs is twolevelset and the other is singlelevelset, then the singlelevelset will need to convertToTwoLevel().
+    /// It's not in parallel and will cost extra large time if the thread_num is large.
+    /// This method will convert all the SingleLevelSet to TwoLevelSet in parallel if the hashsets are not all singlelevel or not all twolevel.
+    static void parallelizeMergePrepare(const std::vector<UniqExactSet *> & data_vec, ThreadPool * thread_pool = nullptr)
```
could we make sure that `ThreadPool` is always not null here and pass it as a reference?

jiebinn replied:
From the `void parallelizeMergePrepare(AggregateDataPtrs & places, ThreadPool & thread_pool) const override` in `AggregateFunctionUniq.h`, the `ThreadPool` is alway not null. The latest code has used the reference instead of a pointer.

nickitat commented on src/AggregateFunctions/UniqExactSet.h:
```diff
@@ -27,6 +27,56 @@ class UniqExactSet
+                    }
+                };
+                for (size_t i = 0; i < std::min<size_t>(thread_pool->getMaxThreads(), single_level_set_num); ++i)
+                    thread_pool->scheduleOrThrowOnError(thread_func);
```
if exception will be thrown here, we will not reach `wait()`. all this code should be wrapped into try-catch that also does `wait` in case of exception (refer to #50590).

jiebinn replied:
Add try catch.

nickitat commented:
would be good to find perf-tests that shows speed-up in the addressed case.

devcrafter commented on src/AggregateFunctions/AggregateFunctionUniq.h:
```diff
@@ -408,6 +409,17 @@ class AggregateFunctionUniq final : public IAggregateFunctionDataHelper<Data, Ag
+
+    void parallelizeMergePrepare(AggregateDataPtrs & places, ThreadPool & thread_pool) const override
+    {
+        std::vector<DataSet *> data_vec;
```
```
data_vec.resize(places.size());
```

jiebinn replied:
I have tried to add the code line `data_vec.resize(places.size());` after the initialization of `std::vector<DataSet *> data_vec;` But the ClickHouse server crashed. I will check that.

jiebinn replied:
Add the resize() for data_vec in parallelizeMergePrepare.

jiebinn commented:
@nickitat @devcrafter Thanks for your kind review. I will think about that and let you know once I finish.

jiebinn commented:
> pls take a look
> 
> ```
> Jun 11 02:52:18 FAILED: src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o 
> Jun 11 02:52:18 /usr/bin/cmake -E __run_co_compile --launcher="prlimit;--as=10000000000;--data=5000000000;--cpu=1000;/usr/bin/sccache" --tidy="/usr/bin/clang-tidy-cache;/usr/bin/clang-tidy-16;--extra-arg-before=--driver-mode=g++" --source=/build/src/Interpreters/examples/two_level_hash_map.cpp -- /usr/bin/clang++-16 --target=x86_64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DANNOYLIB_MULTITHREADED_BUILD -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=10 -DAWS_SDK_VERSION_PATCH=36 -DBOOST_ASIO_HAS_STD_INVOKE_RESULT=1 -DBOOST_ASIO_STANDALONE=1 -DCARES_STATICLIB -DDUMMY_BACKTRACE -DENABLE_ANNOY -DENABLE_MULTITARGET_CODE=1 -DENABLE_QPL_COMPRESSION -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/build/build_docker/includes/configs -I/build/base/glibc-compatibility/memcpy -I/build/src -I/build/build_docker/src -I/build/build_docker/src/Core/include -I/build/base/base/.. -I/build/build_docker/base/base/.. -I/build/contrib/cctz/include -I/build/base/pcg-random/. -I/build/contrib/libfiu/libfiu -I/build/contrib/miniselect/include -I/build/contrib/zstd/lib -I/build/src/Common/mysqlxx/. -I/build/build_docker/rust/skim/include -isystem /build/contrib/llvm-project/libcxx/include -isystem /build/contrib/llvm-project/libcxxabi/include -isystem /build/contrib/libunwind/include -isystem /build/build_docker/contrib/orc/c++/include -isystem /build/contrib/llvm-project/llvm/include -isystem /build/build_docker/contrib/llvm-project/llvm/include -isystem /build/contrib/abseil-cpp -isystem /build/contrib/croaring/cpp -isystem /build/contrib/croaring/include -isystem /build/contrib/sparsehash-c11 -isystem /build/contrib/cityhash102/include -isystem /build/contrib/boost -isystem /build/base/poco/Net/include -isystem /build/base/poco/Foundation/include -isystem /build/base/poco/NetSSL_OpenSSL/include -isystem /build/base/poco/Crypto/include -isystem /build/contrib/boringssl/include -isystem /build/base/poco/Util/include -isystem /build/base/poco/JSON/include -isystem /build/base/poco/XML/include -isystem /build/contrib/replxx/include -isystem /build/contrib/fmtlib-cmake/../fmtlib/include -isystem /build/contrib/magic_enum/include -isystem /build/contrib/double-conversion -isystem /build/contrib/dragonbox/include -isystem /build/contrib/re2 -isystem /build/build_docker/contrib/re2-cmake -isystem /build/contrib/zlib-ng -isystem /build/build_docker/contrib/zlib-ng-cmake -isystem /build/contrib/pdqsort -isystem /build/contrib/xz/src/liblzma/api -isystem /build/contrib/aws/src/aws-cpp-sdk-core/include -isystem /build/build_docker/contrib/aws-cmake/include -isystem /build/contrib/aws/generated/src/aws-cpp-sdk-s3/include -isystem /build/contrib/aws-c-auth/include -isystem /build/contrib/aws-c-common/include -isystem /build/contrib/aws-c-io/include -isystem /build/contrib/aws-crt-cpp/include -isystem /build/contrib/aws-c-mqtt/include -isystem /build/contrib/aws-c-sdkutils/include -isystem /build/contrib/snappy -isystem /build/build_docker/contrib/snappy-cmake -isystem /build/contrib/libbcrypt -isystem /build/contrib/msgpack-c/include -isystem /build/build_docker/contrib/liburing/src/include-compat -isystem /build/build_docker/contrib/liburing/src/include -isystem /build/contrib/liburing/src/include -isystem /build/contrib/fast_float/include -isystem /build/contrib/NuRaft/include -isystem /build/base/poco/MongoDB/include -isystem /build/build_docker/contrib/mariadb-connector-c-cmake/include-public -isystem /build/contrib/mariadb-connector-c/include -isystem /build/contrib/mariadb-connector-c/libmariadb -isystem /build/build_docker/src/Server/grpc_protos -isystem /build/contrib/grpc/include -isystem /build/contrib/c-ares/src/lib -isystem /build/contrib/c-ares/include -isystem /build/contrib/c-ares-cmake/linux -isystem /build/contrib/google-protobuf/src -isystem /build/contrib/s2geometry/src -isystem /build/contrib/AMQP-CPP/include -isystem /build/contrib/AMQP-CPP -isystem /build/contrib/libuv/include -isystem /build/contrib/sqlite-amalgamation -isystem /build/contrib/rocksdb/include -isystem /build/contrib/libpqxx/include -isystem /build/contrib/libpq -isystem /build/contrib/libpq/include -isystem /build/contrib/qpl-cmake -isystem /build/contrib/qpl/include -isystem /build/contrib/idxd-config/accfg -isystem /build/contrib/libstemmer_c/include -isystem /build/contrib/wordnet-blast -isystem /build/contrib/lemmagen-c/include -isystem /build/contrib/ulid-c/include -isystem /build/contrib/consistent-hashing -isystem /build/contrib/annoy/src --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -UNDEBUG -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -g -O0 -g -gdwarf-4  -D_LIBCPP_DEBUG=0 -std=c++2b   -D OS_LINUX -I/build/base -I/build/contrib/magic_enum/include -include /build/src/Core/iostream_debug_helpers.h -Werror -nostdinc++ -MD -MT src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -MF src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o.d -o src/Interpreters/examples/CMakeFiles/two_level_hash_map.dir/two_level_hash_map.cpp.o -c /build/src/Interpreters/examples/two_level_hash_map.cpp
> Jun 11 02:52:18 /build/src/AggregateFunctions/UniquesHashSet.h:369:91: error: unknown type name 'ThreadPool' [clang-diagnostic-error]
> Jun 11 02:52:18     static void parallelizeMergePrepare(const std::vector<UniquesHashSet *> & /*places*/, ThreadPool * /*thread_pool = nullptr*/) {}
> ```

The latest code has involved the head file '<Common/ThreadPool_fwd.h>' in data set, such as 'ThetaSketchData', 'HyperLogLogWithSmallSetOptimization' and 'UniquesHashSet'.

jiebinn commented on src/AggregateFunctions/IAggregateFunction.h:
```diff
@@ -147,6 +148,10 @@ class IAggregateFunction : public std::enable_shared_from_this<IAggregateFunctio
 
+    virtual bool isParallelizeMergePrepareNeeded() const { return false; }
+
+    virtual void parallelizeMergePrepare(AggregateDataPtrs & /*places*/, ThreadPool & /*thread_pool*/) const {}
```
Done. Add the exception in virtual function.

jiebinn commented on src/AggregateFunctions/AggregateFunctionUniq.h:
```diff
@@ -408,6 +409,17 @@ class AggregateFunctionUniq final : public IAggregateFunctionDataHelper<Data, Ag
 
+    bool isParallelizeMergePrepareNeeded() const override { return true;}
+
+    void parallelizeMergePrepare(AggregateDataPtrs & places, ThreadPool & thread_pool) const override
```
Yes. The `is_able_to_parallelize_merge` is unique for each data set. Then I think it could be reused in UniqExactSet. Update it in latest code. 

jiebinn replied:
The `is_able_to_parallelize_merge` could be true or false in different data sets, such as `HyperLogLogWithSmallSetOptimization`, where the merge is not parallel. The optimization patch is only for the data set 'UniqExactSet'. Then only the flag `is_able_to_parallelize_merge` is not enough for `parallelizeMergePrepare`. And we have to add another `is_parallelize_merge_prepare_needed`.

jiebinn commented:
> would be good to find perf-tests that shows speed-up in the addressed case.

I have used the clickbench for the performance test. You mean we might find the "distinct/uniq" operation in tests and check the performance?

jiebinn commented:
Hi @nickitat @devcrafter, thanks for your previous code review. I have updated the patch code and comments according to the review of PR.

jiebinn committed:
Merge pull request #4 from ClickHouse/master

Update to the master

jiebinn committed:
Convert hashSets in parallel before merge

Before merge, if one of the lhs and rhs is singleLevelSet and the other is twoLevelSet,
then the SingleLevelSet will call convertToTwoLevel(). The convert process is not in parallel
and it will cost lots of cycle if it cosume all the singleLevelSet.

The idea of the patch is to convert all the singleLevelSets to twoLevelSets in parallel if
the hashsets are not all singleLevel or not all twoLevel.

I have tested the patch on Intel 2 x 112 vCPUs SPR server with clickbench and latest upstream
ClickHouse.
Q5 has got a big 264% performance improvement and 24 queries have got at least 5% performance
gain. The overall geomean of 43 queries has gained 7.4% more than the base code.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
add resize() for the data_vec in parallelizeMergePrepare()

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn commented:
@nickitat The CI error has succeeded in the latest attempt. But the status doesn't show that.
![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/9e447c01-bd39-4205-852a-b48f2caa56bc)

jiebinn commented:
Update the `clickbench Q5 performance test of core scaling` and `pipeline` figures, which would show how to define the issue and make the `merge` more scalable.

jiebinn commented:
@nickitat @devcrafter @kitaisreal
I have added the performance test for the data set hits_v1, which have 2.1x and 1.8x performance improvement on my machine.

<query>SELECT COUNT(DISTINCT Title) FROM hits_v1 SETTINGS max_threads = 24</query>
<query>SELECT COUNT(DISTINCT Referer) FROM hits_v1 SETTINGS max_threads = 22</query>

jiebinn committed:
Add the performance test prepare_hash_before_merge.xml

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Merge branch 'master' into UniqExactSet

jiebinn commented:
Rename the data set name of performance test from hits_v1 to test.hits to fit the CI rule.

jiebinn committed:
Merge branch 'master' into UniqExactSet

jiebinn committed:
Fit the CI to rename the data set from hits_v1 to test.hits.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Merge branch 'master' into UniqExactSet

nickitat commented on src/AggregateFunctions/ThetaSketchData.h:
```diff
@@ -66,6 +66,8 @@ class ThetaSketchData : private boost::noncopyable
             return 0;
     }
 
+    static void parallelizeMergePrepare(const std::vector<ThetaSketchData *> & /*places*/, ThreadPool & /*thread_pool*/) {}
```
if I'm not missing smth all `parallelizeMergePrepare` could be one of:
* not defined/declared at all
* throw exception
* do meaningful work (empty implementation is not meaningful)

I have some prejudice towards empty methods, what do you think?

jiebinn replied:
Agree. 

- For these empty methods in other datasets, we could add the throw exception. 

- Or we may just delete these empty methods in other datasets. Call the `DataSet::parallelizeMergePrepare()` only when `is_parallelize_merge_prepare_needed` is true.

I would like to suggest the 2nd way.

nickitat commented on src/AggregateFunctions/UniqExactSet.h:
```diff
@@ -28,6 +28,61 @@ class UniqExactSet
+        }
+
+        if (single_level_set_num > 0 && single_level_set_num < data_vec.size())
+            is_convert_needed = true;
+
+        if (is_convert_needed)
```
```suggestion
```

jiebinn committed:
remove the redundant branch in UniqExactSet

Co-authored-by: Nikita Taranov <nickita.taranov@gmail.com>

jiebinn committed:
Remove the empty methods and add throw exception in parallelizeMergePrepare()

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn commented:
The failing check - stress_test (msan) is successful in latest attempt. CI does not update the latest status on the webpage.
![image](https://github.com/ClickHouse/ClickHouse/assets/49742809/9e41df6a-85b6-4004-b8fe-a3de001d892f)
